flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1110,6 +1119,108 @@ def test_no_unmount_if_in_use(self):
         )
 
 
+class BlockDeviceDeployerUpgradeFilesystemTests(SynchronousTestCase):
+    """
+    Tests for automatic upgrade behavior of
+    ``BlockDeviceDeployer.discover_state``
+    """
+    # An id that tests and logging decorators can use.
+    dataset_id = uuid4()
+
+    def _upgrade_ext4_uuid_test(self, mounted):
+        """
+        Test the handling of a Flocker 1.0.0-style filesystem by
+        ``BlockDeviceDeployer.discover_state``.
+
+        :param bool mounted: ``True`` to test the behavior when the filesystem
+            is mounted, ``False`` to tested it unmounted.
+        """
+        deployer = create_blockdevicedeployer(self)
+        api = deployer.block_device_api
+        volume = api.create_volume(
+            dataset_id=self.dataset_id, size=LOOPBACK_MINIMUM_ALLOCATABLE_SIZE
+        )
+        api.attach_volume(
+            blockdevice_id=volume.blockdevice_id,
+            attach_to=api.compute_instance_id(),
+        )
+        device_path = api.get_device_path(volume.blockdevice_id)
+        # Do exactly the same thing that 1.0.0 did so that we're exercising the
+        # upgrade code path we want to be exercising.
+        run_process([
+            b"mkfs", b"-t", b"ext4",
+            # This is ext4 specific, and ensures mke2fs doesn't ask user
+            # interactively about whether they really meant to format whole
+            # device rather than partition.  This stays here forever because
+            # 1.0.0 included it.  It probably doesn't make a difference to the
+            # resulting filesystem state but it's better to be safe.
+            b"-F",
+            device_path.path,
+        ])
+        if mounted:
+            mountpoint = deployer.mountroot.child(bytes(self.dataset_id))
+            mountpoint.makedirs()
+            mount(device_path, mountpoint)
+
+            manifestation = Manifestation(
+                dataset=Dataset(
+                    dataset_id=self.dataset_id,
+                    maximum_size=LOOPBACK_MINIMUM_ALLOCATABLE_SIZE,
+                ), primary=True,
+            )
+            expected_manifestations = {manifestation}
+            expected_nonmanifest_datasets = []
+        else:
+            expected_manifestations = set()
+            expected_nonmanifest_datasets = [self.dataset_id]
+
+        assert_discovered_state(
+            self, deployer,
+            expected_manifestations=expected_manifestations,
+            expected_devices={self.dataset_id: device_path},
+            expected_nonmanifest_datasets=expected_nonmanifest_datasets,
+        )
+        if not mounted:
+            # Only unmounted filesystems are upgraded.
+            self.assertEqual(
+                device_path, get_device_for_dataset_id(self.dataset_id),
+            )
```

Comment:
:question: Should there be an `assertRaises(KeyError,...` here to demonstrate that the filesystem hasn't been upgraded? Or is it simply not possible to run tune2fs on a mounted filesystem?